### PR TITLE
[mlir] [python] Update PyYAML minimum version to 5.4 and limit ml_dtypes to 0.4.0

### PR DESCRIPTION
**Pull request by nhat-nguyen**
PyYAML 5.3.1 has a security vulnerability as described here: https://nvd.nist.gov/vuln/detail/CVE-2020-14343. Update the minimum PyYAML version to 5.4. Also limit ml_dtypes version to 0.4.0.

### CODE REVIEW

File: mlir/python/requirements.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
 numpy>=1.19.5, <=1.26
 pybind11>=2.9.0, <=2.10.3
-PyYAML>=5.3.1, <=6.0.1
-ml_dtypes   # provides several NumPy dtype extensions, including the bf16
\ No newline at end of file
+PyYAML>=5.4, <=6.0.1
```
Comment (sstamenova) [inline mlir/python/requirements.txt:3]: pybind explicitly says 2.9.0, maybe we can explicitly say 5.4.0 for PyYAML for consistency

File: mlir/python/requirements.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
...
+ml_dtypes   # provides several NumPy dtype extensions, including the bf16
```
Comment (sstamenova) [inline mlir/python/requirements.txt:4]: I know your change isn't adding this, but for the same security reason we want to update the PyYAML version, this should have a range or a specific version.

Comment (sstamenova) [reply mlir/python/requirements.txt:4]: @xurui1995 : We can add a reasonable version range here unless you let us know what you would prefer.

Comment (nhat-nguyen) [reply mlir/python/requirements.txt]: ml_dtypes was added around end of May. I see that the latest mldtypes is 0.4.0 which was released on April 1st, so assume that we want at least 0.4.0 for now.

**Review by sstamenova — APPROVED** [review]

File: mlir/python/requirements.txt (hunk 2 above)
Comment (xurui1995) [reply mlir/python/requirements.txt]: At least 0.4.0 seems reasonable，we changed it to at most 0.4.0 now. Any reason? If it’s the requirement of pyyaml I can do a quick test for giving you a minimal version of ml_dtypes

Comment (sstamenova) [reply mlir/python/requirements.txt]: We need an upper bound to prevent potential attacks where new versions have vulnerabilities.

Comment (xurui1995) [reply mlir/python/requirements.txt]: Maybe we can go for `>=0.1.0 <=0.4.0 `, I tried 0.1.0 it still works for the test case in mlir.  For versions under 0.1.0 I think we can ignore them

Comment (nhat-nguyen) [reply mlir/python/requirements.txt]: I have updated to `>=0.1.0, <=0.4.0` as suggested

**Review by xurui1995 — APPROVED** [review]

**Pull request merged**